File: specs/Qowaiv.CodeAnalysis.VisualBasic.Specs/Cases/DecorateFunctions.vb
Imports System
Imports System.Diagnostics.Contracts

Public Class Noncompliant

    Public Function PureFunction() As Integer ' Noncompliant {{Decorate this method with a [Pure] or [Impure] attribute.}}
        Return 42
    End Function

    Public Function PureFunction(scale) As Integer ' Noncompliant
        Return scale * 42
    End Function

    <OtherAttribute>
    Public Function DecoratedOtherwise() As Integer ' Noncompliant@-1
        Return 42
    End Function
End Class

Public Class Compliant
    <Pure>
    Public Function PureFunction() As Integer
        Return 42
    End Function

    <Obsolete>
    Public Function Obsolete() As Integer ' Compliant {{Obsolete methods are ignored.}}
        Return 0
    End Function

    Public Sub Void() 'Compliant {{Void methods are impure per definition.}}
    End Sub

    Public Function Scope() As IDisposable ' Compliant {{Disposable methods are expected To be impure.}}
        Return Nothing
    End Function

    Public Function TryParse(str As String, ByRef result As Object) As Boolean ' Compliant {{Methods With out parameters are expected To be impure.}}
        If str Is Nothing Then
            result = Nothing
            Return False
        Else
            result = str
            Return True
        End If
    End Function

    <Impure>
    Public Function ImpureMethod(input As Integer) As Integer ' Compliant {{Decorated With something derived from an ImpureAttribute.}}
        Return 69
    End Function

    <FluenSyntax>
    Public Function FluentSyntax() As Compliant ' Compliant {{Decorated With something derived from an ImpureAttribute.}}
        Return Me
    End Function
End Class

Public Class OtherAttribute
    Inherits Attribute
End Class
Public Class ImpureAttribute
    Inherits Attribute
End Class
Public Class FluenSyntaxAttribute
    Inherits ImpureAttribute
End Class

File: specs/Qowaiv.CodeAnalysis.CSharp.Specs/Cases/DecorateFunctions.cs
using System;
using System.Diagnostics.Contracts;
using FluentAssertions;

public class Noncompliant
{
    public int PureFunction() => 42; // Noncompliant {{Decorate this method with a [Pure] or [Impure] attribute.}}
//  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    public int PureFunction(int scale) => scale * 42; // Noncompliant

    [Other]
    public int DecoratedOtherwise() => 42; // Noncompliant@-1
}

public class Compliant
{
    [Pure]
    public int PureFunction() => 42;

    [Obsolete]
    public int Obsolete() => 0; // Compliant {{Obsolete methods are ignored.}}

    public void Void() { } // Compliant {{Void methods are impure per definition.}}

    public IDisposable Scope() => null; // Compliant {{Disposable methods are expected to be impure.}}

    public bool TryParse(string str, out object result) // Compliant {{Methods with out parameters are expected to be impure.}}
    {
        if (str is null)
        {
            result = default;
            return false;
        }
        else
        {
            result = str;
            return true;
        }
    }

    public int GuardPositive(int number) => number; // Compliant {{Guarding is expected to be impure.}}

    [Impure]
    public int ImpureMethod(int input) => 69; // Compliant {{Decorated with something derived from an ImpureAttribute.}}

    [FluenSyntax]
    public Compliant FluentSyntax() => this; // Compliant {{Decorated with something derived from an ImpureAttribute.}}

    [CustomAssertion]
    public T SomeAssertion<T>(T subject) => subject; // Compliant {{FluentAssertions custom assertions are expected to be impure.}}
}

public class Guard
{
    public T NotNull<T>(T obj) => obj; // Compliant {{Methods on a Guard class are expected to be impure.}}
}

public class OtherAttribute : Attribute { }
public class ImpureAttribute : Attribute { }
public class FluenSyntaxAttribute : ImpureAttribute { }
